app/models/user.rb

```diff
@@ -80,8 +80,60 @@ class User < ApplicationRecord
 
   attribute :renewal_request, :string
 
-  attribute :created_at, :datetime, precision: 6
-  attribute :updated_at, :datetime, precision: 6
-
   validates :email, presence: true
+
+  # Finds, creates, or updates user from userinfo
+  # Find in case of user with existing token matching userinfo["sub"]
+  # Create in case of no token or email matching in userinfo
+  # Update in case of matching email to userinfo["email"] but no token set
+  # TODO: Add relevant security log tracking here?
+  def self.user_from_userinfo(userinfo)
+    email = userinfo[0]["email"]
+    token = userinfo[0]["sub"]
+
+    if (user = find_by(token:))
+      user
+    elsif (user = find_by(email:))
+      update_admin_added_user(user, userinfo)
+    else
+      create_user_from_userinfo(userinfo)
+    end
+  end
+
+  def self.update_admin_added_user(user, userinfo)
+    update(user.id, { token: userinfo[0]["sub"] })
+  end
+
+  def self.create_user_from_userinfo(userinfo)
+    email = userinfo[0]["email"]
+    token = userinfo[0]["sub"]
+
+    default_role_and_status = default_role_and_status_for_email(email)
+    default_role = default_role_and_status[0]
+    default_status = default_role_and_status[1]
+
+    create({
+             email:,
+             role: default_role,
+             token:,
+             terms_of_use: nil,
+             privacy_guidelines: nil,
+             status: default_status
+           })
+  end
+
+  def self.default_role_and_status_for_email(email)
+    if default_challenge_manager?(email)
+      %w[challenge_manager pending]
+    else
+      %w[solver active]
+    end
+  end
+
+  def self.default_challenge_manager?(email)
+    escaped_gov_tld = Regexp.escape('.gov')
+    matching_gov_string = ".*#{escaped_gov_tld}$"
+    gov_regex = Regexp.new(matching_gov_string)
+    gov_regex.match?(email)
```

Comment:
I think we're supposed to [include `.mil` emails](https://github.com/GSA/Challenge_gov/issues/1390) (should add a test too)
```suggestion
    /\.(gov|mil)$/.match?(email)
```